# Redondeo

### Ejercicio 1

 (i) Si tuviéramos un número real positivo x (es decir, con precisión infinita), ¿cómo podríamos encontrar △(x) y ▽(x)?

**R:** Necesitaríamos conocer la precisión de nuestro sistema de punto flotante. Teniendo esta en mente podríamos truncar hasta el número de cifras significativas permitidas. Al número truncado resultante vemos si pertenece al sistema de punto flotante dado, si no pertenece incrementamos su valor hasta el siguiente número que si pertenezca al sistema para el caso de redondeo hacia arriba o disminuimos su valor al siguiente número para el caso del redondeo hacia abajo.

(ii) Encuentra △(0.1) y ▽(0.1) para aritmética flotante de IEEE. ¿En cuánto difieren? ¿Qué podemos decir sobre el error de redondeo?

**R:** Primero algunas obervaciones interesantes:

In [2]:
bits(0.1)

"0011111110111001100110011001100110011001100110011001100110011010"

Como habíamos visto desde la práctica anterior, el real 0.1 tiene una representació periódica en el sistema binario. Esto quiere decir que la cadena de 110011001100... se repetirá de manera infinita. En la instrucción anterior vemos como el número se guarda con precisión simple. Se trunca la cadena en una secuencia 110011001 y se le suma 1 obteniendo como resultado 110011010 . Esto quiere decir que se esta llevando a cabo un Redondeo hacia arriba. Para el caso de redonde hacia abajo bastaría truncar la cadena y proporcionar el número representado por 110011001 . 

Utilizando precisión extendida tenemos el siguiente resultado:

In [1]:
big(0.1)

1.000000000000000055511151231257827021181583404541015625e-01 with 256 bits of precision

Aqui vemos claramente que el sistema de punto flotante efectivamente no guarda el número 0.1 sino el 1.000000000000000055511151231257827021181583404541015625e-01, esto quiere decir que hasta dígito número 17 se ve afectado el resultado. Esto coicide con el siguiente experimento para cuando utilizamos precisión simple.

In [2]:
function RedondeoUp(x)
    y=0.1
    z=1
    while z>x
        y=y/10
        z=y+x
    end
    println(y)#Imprime el número más grande posible que al sumar no afecte el resultado
    println(x+y)#Imprime la suma de la entrada y el número anterior
    println(y*10)#Imprime el número más pequeño posible que al sumar si afecta el resultado
    println(x+y*10)
end

RedondeoUp (generic function with 1 method)

In [3]:
RedondeoUp(0.1)

1.0e-18
0.1
1.0e-17
0.10000000000000002


El resultado de la función anterior nos dice que el número más pequeño que podemos sumar a un flotante en preción simple es del orden de 1.0e-17, y cualquier número más chico no lo alcanzará a sumar. Adicionalmente verificamos que se aplica el redondeo hacia arriba pues no le es posible guardar el número 0.10000000000000001.

Lo que podemos decir del error de redondeo es que este es del orden de 1.0e-17

### Ejercicio 2

Haz de nuevo el ejercicio anterior para x=1.1 y para x=10.1.

**R:** Vemos los siguientes resultados:

In [1]:
big(1.1)

1.100000000000000088817841970012523233890533447265625e+00 with 256 bits of precision

In [1]:
big(10.1)

1.00999999999999996447286321199499070644378662109375e+01 with 256 bits of precision

Notamos que conforme utilizamos más bits en la parte entera vamos perdiendo precisión en la parte fraccionaria. 

### Ejercicio 3

¿Qué pasa con △(x) y ▽(x) si x∈F∗?

**R:** Por la definición misma de redondeo dada es claro que obtendríamos el mismo resultado x para ambos redondeos. Esta es una propiedad del mapeo del redondeo pues establece que todo número flotante es fijo bajo el redondeo, claramente un número real que ya es representable en flotante no requiere de algún redondeo (Tucker).

### Ejercicio 4

¿Cuál es la relación entre △(−x) y ▽(x)?

**R:** Satisfacen una relación antisimétrica pues: △(−x)= -▽(x)

### Ejercicio 5

 Encuentra unos ejemplos de pares de números x,y∈F tal que x⊕y∉F. (Aquí, F denota a los flotantes de doble precisión de IEEE, y ⊕ es alguna operación aritmética entre x y y.)

In [1]:
x=big(0.125)#Tiene representación flotante

1.25e-01 with 256 bits of precision

In [2]:
y=big(0.0001220703125)#Tiene representación flotante

1.220703125e-04 with 256 bits of precision

In [7]:
x+y#Tiene representación flotante

1.251220703125e-01 with 256 bits of precision

### Ejercicio 6

¿Qué podemos hacer al respecto?

### Ejercicio 7

En los reales tenemos que, si se cumple x+y=x+y′, entonces y=y′. ¿Se cumple esto entre los números de punto flotante? Si tu respuesta es no, da un ejemplo.

**R:** Esto no siempre se cumple, a continuación presento un ejemplo utilizando precisión simple:

In [5]:
a=1
b=0.000000000000000001
c=0

1.0

In [6]:
a+b

1.0

In [7]:
a+c

1

In [8]:
c==b

false

### Ejercicio 8

 Analiza el caso de iterar el mapeo f:[0,1]→[0,1] dado por f(x)=3x mod 1, con la condición inicial x0=110:
 
    ¿Qué pasa analíticamente?
    ¿Qué pasa numéricamente?
    ¿Qué pasa si consideras una condición inicial x0 arbitraria?


### Ejercicio 9

 Calcula S numéricamente de manera ingenua. 

In [8]:
S=0
for n=1:1000000
    S=S+1/n^2
end
println("Utilizando 1000000 iteraciones")
println("El valor de S utilizando un ciclo es de $S")
error=abs(S-pi^2/6)#Sabemos analíticamente que la solucion es pi^2/6
println("El error es de $error")


Utilizando 1000000 iteraciones
El valor de S utilizando un ciclo es de 1.64493306684877
El error es de 9.999994563525405e-7


Sea la cola de la suma TN:=∑∞n=N+11n2. Utiliza un argumento geométrico para mostrar que
$∫∞N+11x2dx<TN<∫∞N+11(x−1)2dx$,

y así encuentra cotas para TN.